pkg/apply/apply.go

```diff
@@ -0,0 +1,63 @@
+package apply
+
+import (
+	"context"
+	"fmt"
+	"github.com/pkg/errors"
+	"log"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	uns "k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	k8sclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func getExistingObject(ctx context.Context, client k8sclient.Client, obj *uns.Unstructured) (*uns.Unstructured, string, error) {
+	name := obj.GetName()
+	namespace := obj.GetNamespace()
+	if name == "" {
+		return nil, "", errors.Errorf("Object %s has no name", obj.GroupVersionKind().String())
+	}
+	gvk := obj.GroupVersionKind()
+	// used for logging and errors
+	objDesc := fmt.Sprintf("(%s) %s/%s", gvk.String(), namespace, name)
+	log.Printf("reconciling %s", objDesc)
+
+	// Get existing
+	existing := &uns.Unstructured{}
+	existing.SetGroupVersionKind(gvk)
+	err := client.Get(ctx, types.NamespacedName{Name: obj.GetName(), Namespace: obj.GetNamespace()}, existing)
+	return existing, objDesc, err
+}
+
+// ApplyObject applies the desired object against the apiserver,
+// merging it with any existing objects if already present.
+func ApplyObject(ctx context.Context, client k8sclient.Client, obj *uns.Unstructured) error {
+	existing, objDesc, err := getExistingObject(ctx, client, obj)
+
+	if err != nil && apierrors.IsNotFound(err) {
+		log.Printf("does not exist, creating %s", objDesc)
+		err := client.Create(ctx, obj)
+		if err != nil {
+			return err
+		}
+		log.Printf("successfully created %s", objDesc)
+		return nil
+	}
+
+	// Merge the desired object with what actually exists
+	if err := MergeObjectForUpdate(existing, obj); err != nil {
```

Comment:
After a quick glance, I'm wondering if merge.go is necessary or if we can get rid of it.

I suppose that `mergeMetadataForUpdate` is used so that you can do `equality.Semantic.DeepEqual` --- but you could also just DeepEqual the specs and status, couldn't you?

I know that this is generic code, but it's for Deployment, ServiceAccount and Service. All things that we are not going to manipulate. It seems to me that the entire merge part might be redundant?

But I very likely am missing something :-)

Reply:
the idea from merge it walks the existing objects and merge them to the new object so at the end we really just update the delta so you don't lose prev configured objects.
I agree with you not really needed for our ds deployement I will take it off and bring it back if needed